### PR TITLE
Implement benchmarks for dmsg using map and array

### PR DESCRIPTION
**Pull request by nkryuchkov**
WIP for #8.

 Changes:	
- Implemented  `ClientConn`, `AckWaiter` and `ServerConn` using map instead of array
- Implemented benchmarks for `ClientConn`, `AckWaiter` and `ServerConn` using maps
- Implemented benchmarks for `ClientConn`, `AckWaiter` and `ServerConn` using arrays

How to test this PR:
`go test -bench=. -benchmem`

Results:
image

Conclusion:
It looks like the array is slightly faster than the map at accessing, however a lot slowlier at creation. Moreover, every instance `ClientConn`, `AckWaiter` and `ServerConn` takes about 500 KB memory for storing transports/connections. For example, 1000 `ClientConn`s + 1000 `AckWaiter`s + 1000 `ServerConn`s would take about 1,5 GB memory.

### CODE REVIEW

**Comment by evanlinjin** [comment]
@nkryuchkov We also have [`AckWaiter`](https://github.com/skycoin/dmsg/blob/e69f75132be9ad1070b8e3d66f9e13a37490e8af/ioutil/ack_waiter.go#L33) structure that uses an array.

Also, I would be curious on the benchmark of `Transport.Read`/`Transport.Write` operations (in total).

**Comment by nkryuchkov** [comment]
@evanlinjin Yeah, I meant that `Transport` uses `AckWaiter` that uses an array. I'll fix this in the description.

I pushed benchmarks for `Transport.Read`/`Transport.Write`. It seems that the difference is very small:
image

**Comment by evanlinjin** [comment]
@nkryuchkov I would like to keep these benchmarks. However, it will be good if we can get rid of the `client_map.go`, `transport_map.go`, etc. and associated structures first.

**Comment by nkryuchkov** [comment]
@evanlinjin Do you mean keeping the map benchmarks only? If we get rid of those files and associated structures, we would have only one implementation. 

Probably, I didn't fully understand what you meant.

**Comment by evanlinjin** [comment]
One implementation is fine. We are using maps for both the implementations currently.

**Comment by nkryuchkov** [comment]
@evanlinjin Okay, I'll leave only one implementation with its benchmarks

**Comment by nkryuchkov** [comment]
@evanlinjin Got rid of them

File: transport_test.go [hunk 1]
```diff
@@ -47,3 +52,100 @@ func TestTransport_close(t *testing.T) {
...
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	const bufSize = 50000
+	buf := make([]byte, bufSize)
+	go func() {
+		for {
+			initTr.Read(buf) // nolint:errcheck
+			time.Sleep(10 * time.Microsecond)
```
Comment (evanlinjin) [inline transport_test.go]: Why is this `Sleep` needed?

**Review by evanlinjin — APPROVED** [review]
Good work! Just a single query.

File: transport_test.go (hunk 1 above)
Comment (nkryuchkov) [reply transport_test.go]: @evanlinjin I thought I might give some rest to CPU. I removed this line and run the tests again, saw no significant change. It probably may be removed as well as the `Sleep` in `BenchmarkTransport_Read`.

Comment (nkryuchkov) [reply transport_test.go]: @evanlinjin Removed

**Pull request merged**